content/docs/user-guide/data-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,221 @@
+the output of one becomes the input of another, and so on. Technically, this is
+called a _dependency graph_ (DAG).
+
+## Directed Acyclic Graph (DAG)
+
+DVC represents a pipeline internally as a _graph_ where the nodes are stages and
+the edges are _directed_ dependencies (e.g. A before B). In order for DVC to
+execute the pipeline reliably, its topology should be _acyclic_ -- because
+executing cycles (e.g. A -> B -> C -> A ...) would continue indefinitely. [More
+about DAGs].
+
+<admon type="info">
+
+Stage execution will be determined entirely by the DAG, not by the order in
+which stages are found in `dvc.yaml`.
+
+</admon>
+
+Use `dvc dag` to visualize (or export) pipeline DAGs.
+
+[more about dags]: https://en.wikipedia.org/wiki/Directed_acyclic_graph
+
+## Stages
+
+<admon type="tip">
+
+See the full [specification] of stage entries.
+
+[specification]: /doc/user-guide/project-structure/dvcyaml-files#stage-entries
+
+</admon>
+
+Each stage wraps around an executable shell [command](#stage-commands) and
+specifies the necessary inputs as well as expected outputs (if any). Let's look
+at an example that depends on a script file it runs and on a raw data directory
+(ideally [tracked by DVC] already):
+
+```yaml
+stages:
+  extract:
+    cmd: source src/cleanup.sh
+    deps:
+      - src/cleanup.sh
+      - data/raw
+    outs:
+      - data/clean.csv
+```
+
+<admon type="info">
+
+We use [GNU/Linux](https://www.gnu.org/software/software.html) in these
+examples, but Windows or other shells can be used too.
+
+</admon>
+
+Besides writing `dvc.yaml` files manually (recommended), you can also create
+stages with `dvc stage add` -- a limited command-line interface to setup
+pipelines. Let's add another stage this way and look at the resulting `dvc.yaml`
+file:
+
+```dvc
+$ dvc stage add --name transform \
+                --deps src/process.py \
+                --deps data/clean.csv \
+                --outs data/features.dat \
+                python src/process.py data/clean.csv
+```
+
+```yaml
+stages:
+  extract:
+    ...
+    outs:
+      - data/clean.csv
+  transform:
+    cmd: python src/process.py data/clean.csv
+    deps:
+      - src/preprocess.py
+      - data/clean.csv
+    outs:
+      - data/features.dat
+```
+
+<admon type="tip">
+
+One advantage of using `dvc stage add` is that it will verify the validity of
+the arguments provided (otherwise stage definition won't be checked until
+execution). A disadvantage is that some advanced features such as [templating]
+are not available this way.
+
+[templating]: /doc/user-guide/project-structure/pipelines-files#templating
+
+</admon>
+
+Notice that the new `transform` stage [depends](#simple-dependencies) on the
+output from stage `extract` (`data/clean.csv`), forming the
+[DAG](#directed-acyclic-graph-dag).
+
+[tracked by dvc]: /doc/start/data-management
+
+## Stage commands
+
+The command(s) defined in the `stages` (`cmd` field) can be anything your system
```

Comment:
I think these details if we remove them from cmd ref should go into DVC.yaml lang spec (Project Structure)

Reply:
I removed them from the cmd ref because it was repeated in 3 pages (`run`, `stage add`, and `exp init`). I agree that it makes more sense to put it in the dvc.yaml schema ref. Done